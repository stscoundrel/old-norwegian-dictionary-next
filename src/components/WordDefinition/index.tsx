import { capitalize } from 'lib/utils/strings'
import { lettersToShortTwigRunes } from 'younger-futhark'
import { Abbreviation, addAbbreviationsToContent } from 'lib/services/abbreviations'
import { DictionaryEntry } from 'lib/models/dictionary'
import Abbreviations from 'components/Abbreviations'
import styles from './WordDefinition.module.scss'

interface WordDefinitionProps{
  data: DictionaryEntry,
  abbreviations: Abbreviation[]
}

export default function WordDefinition({ data, abbreviations }: WordDefinitionProps) {
  const { word, partOfSpeech, definition } = data

  return (
    <article className={styles.section}>
      <header>
        <h1 lang="non">{capitalize(word)}</h1>

        <small className={styles.subHeading}>
          Old Norwegian Dictionary - {word.toLowerCase()}
        </small>
        <p>Meaning of Old Norwegian word <em>&quot;{word}&quot;</em> in Norwegian.</p>
        <p>As defined by the Johan Fritzer&apos;s Old Norwegian dictionary:</p>
      </header>

      <dl className={styles.definitionList}>
        <dt><strong>{word}</strong></dt>
        <dd
          lang="nor"
          className={styles.itemDescription}
          dangerouslySetInnerHTML={{
            __html: addAbbreviationsToContent(definition, abbreviations),
          } }
        ></dd>
      </dl>

      <p><strong>Part of speech:</strong> <em>{partOfSpeech}</em></p>

      <p>Possible runic inscription in <em>Younger Futhark:</em>
        <span className={styles.rune}>{ lettersToShortTwigRunes(word) }</span>
      </p>

      <Abbreviations abbreviations={abbreviations} />
    </article>
  )
}
