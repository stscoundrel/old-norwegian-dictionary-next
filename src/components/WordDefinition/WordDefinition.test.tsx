import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import { DictionarySource } from 'scandinavian-dictionary-crosslinker'
import WordDefinition from './index'

const word = {
  word: 'þrályndi',
  partOfSpeech: 'n',
  definition: 'þrályndi, n. d. s.; hélt honum aptr fornsiðvenja ok mykyt þrályndi Barl. 12524;Fgrdl. 5816.',
  slug: 'thralyndi',
}

const wordWithOlderSpelling = {
  word: 'völlr',
  partOfSpeech: 'm',
  definition: 'völlr, m. (Gen. vallar, n. Pl. vellir, A. Pl.völlu) fast og jevn Jordvold',
  slug: 'vollr',
}

const abbreviations = [
  {
    abbreviation: 'n.',
    explanation: 'Neutrum.',
  },
  {
    abbreviation: 's.',
    explanation: 'substantiv.',
  },
]

const crosslinks = [
  {
    url: 'https://old-icelandic.vercel.app/word/fadir',
    source: DictionarySource.OldIcelandic,
  },
  {
    url: 'https://old-swedish-dictionary.vercel.app/word/fadhir',
    source: DictionarySource.OldSwedish,
  },
]

describe('WordDefinition component', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<WordDefinition data={word} abbreviations={abbreviations} crosslinks={crosslinks}/>)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <WordDefinition data={word} abbreviations={abbreviations} crosslinks={crosslinks} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Matches snapshot (older spelling variant)', () => {
    const tree = renderer.create(
      <WordDefinition
        data={wordWithOlderSpelling}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct label', () => {
    const tree = renderer.create(
      <WordDefinition data={word} abbreviations={abbreviations} crosslinks={crosslinks} />,
    )
    const { root } = tree

    expect(root.findByType('h1').children).toEqual(['Þrályndi'])
  })
})
