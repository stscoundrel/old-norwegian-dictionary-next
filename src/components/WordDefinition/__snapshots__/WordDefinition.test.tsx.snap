// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordDefinition component Matches snapshot 1`] = `
<article
  className="section"
>
  <header>
    <h1
      lang="non"
    >
      Þrályndi
    </h1>
    <small
      className="subHeading"
    >
      Old Norwegian Dictionary - 
      þrályndi
    </small>
    <p>
      Meaning of Old Norwegian word 
      <em>
        "
        þrályndi
        "
      </em>
    </p>
  </header>
  <dl
    className="definitionList"
  >
    <dt>
      <strong>
        þrályndi
      </strong>
    </dt>
    <dd
      className="itemDescription"
      dangerouslySetInnerHTML={
        Object {
          "__html": "þrályndi, <abbr title=\\"Neutrum.\\">n.</abbr> d. <abbr title=\\"substantiv.\\">s.</abbr>; hélt honum aptr fornsiðvenja ok mykyt þrályndi Barl. 12524;Fgrdl. 5816.",
        }
      }
      lang="nor"
    />
  </dl>
  <p>
    <strong>
      Part of speech:
    </strong>
     
    <em>
      n
    </em>
  </p>
  <p>
    Possible runic inscription in 
    <em>
      Younger Futhark:
    </em>
    <span
      className="rune"
    >
      ᚦᚱᛆᛚᚢᚿᛐᛁ
    </span>
  </p>
  <div
    className="abbreviations"
  >
    <div
      className="column"
    >
      <h4>
        Abbreviations used:
      </h4>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            n.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          Neutrum.
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            s.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          substantiv.
        </dd>
      </dl>
    </div>
  </div>
</article>
`;
